src/tests/testClpMap.cc

```diff
@@ -0,0 +1,129 @@
+testClpMap::addData(testMap &m, int numElems, int base, testMap::Ttl ttl)
+{
+    for (int j = base; j < base + numElems; ++j ) {
+        CPPUNIT_ASSERT_EQUAL(true, m.add(std::to_string(j), j, ttl));
+    }
+}
+
+
+void
+testClpMap::setUp()
+{
+    squid_curtime = time(nullptr);
+}
+
+void
+testClpMap::PutGetDelete()
+{
+    testMap m(1024);
+    addData(m, 10);
+    CPPUNIT_ASSERT_EQUAL(static_cast<const int *>(nullptr), m.get("notthere"));
+    CPPUNIT_ASSERT_EQUAL(1, *(m.get("1")));
+    CPPUNIT_ASSERT_EQUAL(9, *(m.get("9")));
+    m.add("1", 99);
+    CPPUNIT_ASSERT_EQUAL(99, *(m.get("1")));
+    m.del("1");
+    CPPUNIT_ASSERT_EQUAL(static_cast<const int *>(nullptr), m.get("1"));
+}
+
+void testClpMap::Entries()
+{
+    {
+        testMap m(10*1024*1024, 10);
+        addData(m, 10, 10);
+        CPPUNIT_ASSERT_EQUAL(static_cast<size_t>(10), m.entries());
+        m.add("foo", 0, 10);
+        CPPUNIT_ASSERT_EQUAL(static_cast<size_t>(11), m.entries());
+    }
+    {
+        testMap m(1024, 5);
+        addData(m, 1000);
+        CPPUNIT_ASSERT(m.entries() < 1000);
+    }
+}
+
+void
+testClpMap::testMemoryCounter()
+{
+    CPPUNIT_ASSERT_EQUAL(sizeof(int), static_cast<size_t>(DefaultMemoryUsage(int())));
+    CPPUNIT_ASSERT_EQUAL(sizeof(int32_t), static_cast<size_t>(DefaultMemoryUsage(int32_t())));
+    CPPUNIT_ASSERT_EQUAL(sizeof(int64_t), static_cast<size_t>(DefaultMemoryUsage(int64_t())));
+    CPPUNIT_ASSERT_EQUAL(sizeof(char), static_cast<size_t>(DefaultMemoryUsage(char())));
+    char str[10];
+    CPPUNIT_ASSERT_EQUAL(sizeof(str), static_cast<size_t>(DefaultMemoryUsage(str)));
+    CPPUNIT_ASSERT_EQUAL(sizeof(std::string), static_cast<size_t>(DefaultMemoryUsage(std::string())));
+}
+
+void
+testClpMap::testConstructor()
+{
+    testMap nilA(0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilA.entries());
+
+    testMap nilB(0, 0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilB.entries());
+
+    testMap emptyC(1);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyC.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyC.entries());
+
+    testMap emptyD(1024);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyD.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyD.entries());
+}
+
+void
+testClpMap::SetMemLimit()
+{
+    testMap m(2048);
+    addData(m, 1000);
+    auto entriesBefore = m.entries();
+    m.setMemLimit(1024);
+    CPPUNIT_ASSERT(entriesBefore > m.entries());
```

Comment:
I agree that we should test what Amos is suggesting to test. However, please note that memoryUsed() for a one-item map is not memPerEntry (i.e. not memory used per each entry) because the map is not guaranteed to use the same amount of memory for each entry. There should be a way to test all of the above _without_ making this assumption.

Reply:
I would argue that ClpMap would be more useful if it didn't try to be too clever about the storage it uses but implemented limits to the number of entries instead, or at least in addition to the current limits.
In fact, I'd rather suggest to build a strategy around that: we test the API we have, extend the API to limit also by number of entries, and test that one instead of kludging around the heuristics about an entry's size

Reply:
See https://github.com/squid-cache/squid/pull/1262#discussion_r1112021206 for the API discussion. It applies to the "support two kind of limits at the same time" suggestion as well.

Reply:
> However, please note that memoryUsed() for a one-item map is not memPerEntry

Ah right. The key length is variable for this test code.  That could be avoided for test calculations by `addData()` ensuring a fixed-length key string (i.e. 5 bytes "0"-padded).

@rousskov, is that the only variability you see with my suggested test?

Reply:
> @rousskov, is that the only variability you see with my suggested test?

Once I found memPerEntry, I only looked through the high-level C++ comments describing the new tests, not the suggested code itself (because the code depended heavily on the problematic memPerEntry). I plan to revisit this code when this change request is implemented (if it has not been implemented already in the code I reviewed).

Overall, I strongly recommend testing the overall map functionality _without_ assuming exactly how big each entry is, what is the size of various housekeeping information that the map might account for, etc. Otherwise, we risk re-implementing current ClpMap internal code assumptions/decisions in tests, which only makes things worse overall, without significantly increasing the probability of catching implementation bugs.

For example, when we shrink capacity after overfilling the map, we may not be able to predict exactly how many entries will be purged and that is perfectly fine. Instead, we can predict (and should validate) that:

* some entries were purged
* purged entries were purged in LRU order
* some entries are preserved (until we shrink the capacity to zero -- then zero entries should be kept!)
* resulting memory usage (as reported by the map) is at or below the new limit
* adding new entries is still possible

We could also test that, after shrinking the memory limit by half, adding a single entry will purge at least one more entry, but I would _not_ do that because it starts to assume a lot about internal ClpMap decisions. It is possible to imagine a perfectly valid ClpMap implementation that would not pass such a test in various corner cases. We could test that adding, say, 10 new entries after shrinking does purge at least one old entry.

Reply:
> > However, please note that memoryUsed() for a one-item map is not memPerEntry ... because the map is not guaranteed to use the same amount of memory for each entry
 
> Ah right. The key length is variable for this test code.

Actually, the primary issue here is not so much the changes in the key length. The issue is that the amount of internal housekeeping information[^1] that the map uses to store items may change as more items are added. For example, the map may allocate a second "bucket" when the first "bucket" gets full (where "bucket" is some undefined internal storage implementation thing). Or create an index but only when the number of entries exceed 100. My point was that we should not (and do not need to) use memPerEntry for this kind of tests.

We could add tests where average memPerEntry or similar stats might be useful. For example, we can test that if the map has more than 10'000 entries, then the housekeeping information does not consume more than 10% of raw/pure entry sizes. However, I doubt we should spend Francesco's time on such tests, in this PR! ;-)

[^1]:  Housekeeping structures do (or should and, hence, eventually will) affect the total reported map memory usage.

Reply:
Oh. That level of housekeeping is too deep for the current ClpMap accounting to even know about.

@kinkie, if you want to defer this change to a future PR I can take a shot at adding a fixed version of my sketch later.

Reply:
> Oh. That level of housekeeping is too deep for the current ClpMap accounting to even know about.

ClpMap already accounts for some housekeeping information at that level. See MemoryCountedFor() where, out of four counted memory consumers, two are about housekeeping overheads.

Reply:
> For example, when we shrink capacity after overfilling the map, we may not be able to predict exactly how many entries will be purged and that is perfectly fine. Instead, we can predict (and should validate) that:
> 
> * some entries were purged

Isn't this tested in testSetMemLimit ?

> * purged entries were purged in LRU order (in aggregate; i.e. any entry that was kept was more recently "used" than any non-expired entry that was purged)

This is a great idea. Added testClpMap::testPurgeIsLRU

> * some entries are preserved (until we shrink the capacity to zero -- then zero entries should be kept!)

extended testClpMap::testSetMemLimit to check this.
It's a bit kludgy, and it requires making TestMap::MemoryCountedFor public.
By the way, why does it return an optional? I can't see in the code any path that could make it not return a value.

> * resulting memory usage (as reported by the map) is at or below the new limit
> * adding new entries is still possible

> We could also test that, after shrinking the memory limit by half, adding a single entry will purge at least one more entry, but I would _not_ do that because it starts to assume a lot about internal ClpMap decisions. It is possible to imagine a perfectly valid ClpMap implementation that would not pass such a test in various corner cases. We could test that adding, say, 10 new entries after shrinking does purge at least one old entry.

I don't think these cases will actually matter in practice. We could argue that even setting a ClpMap to be capped at 0 size might make no sense in practical terms because that's a degenerate behaviour; but in the end it's okay to add tests for such contingencies, if we stop somewhere sensible

Reply:
FYI, for some of the cache types this is intended to replace (eg helper responses) there are valid situations where admin set capacity to 0 bytes. Having the ClpMap handle that saves callers individually duplicating all the logic testing if cache is disabled.

Reply:
> FYI, for some of the cache types this is intended to replace (eg helper responses) there are valid situations where admin set capacity to 0 bytes. Having the ClpMap handle that saves callers individually duplicating all the logic testing if cache is disabled.

Got you.
The only thing that makes life complicated IMO for callers is that add() returns false if the entry could not be added because memlimit is such that the map can't even store one single entry. This means that either add() return value should be ignored, or that callers have to special case this anyway. Have you considered this?

Reply:
> Francesco:  By the way, why does TestMap::MemoryCountedFor() return an optional? I can't see in the code any path that could make it not return a value.

Adding an entry with a ULLONG_MAX-byte key[^1] is one of the use cases where MemoryCountedFor() cannot return an accurate answer and, hence, should not return any answer. This is one of the missing test cases, BTW.

The only alternative is an assertion, but since ClpMap is essentially a cache, it feels wrong to kill Squid just because we cannot cache.

[^1]: Which might happen due to `-1` to `unsigned` conversion bugs in caller's implementation of the MemoryUsedBy() API. Similar Squid bugs have been discovered in the past, of course, and I am pretty sure there will be more.


> either add() return value should be ignored, or that callers have to special case this anyway

The API accommodates callers that want to debugs() the outcome or update some indexes/stats. And even among those callers that pay attention to the return value, many will not have to special-case it. Many examples of such paying-attention not-special-casing (in the true/core meaning of that phrase) callers are added in this PR.

Reply:
> FYI, for some of the cache types this is intended to replace (eg helper responses) there are valid situations where admin set capacity to 0 bytes.

Agreed. I would even assert that a zero-capacity (or equivalent) cache should be "valid" in the _majority_ of use cases, especially where the cache capacity is admin-configurable or dynamically computed from environment characteristics outside code control.